### ISSUE TITLE
created is always added to the signature 

### ISSUE THREAD

**Issue by mjeffrey** [tomitribe/http-signatures-java#50]
Hi 
Not sure if this is a bug but after an upgrade to 1.7 I see 'created' added systematically.

If created is there I would have expected  `(created)` in the headers?

I am concerned that when the signature is generated the verifier (of thirdparty banks) may not understand it (the current version works OK) but maybe other implementations could fail.

I tried to disable it but there does not seem to be a way to do this (it is set in the Signer.sign method)
```
public Signature sign(final String method, final String uri, final Map<String, String> headers) throws IOException {
       final Long created = System.currentTimeMillis();
```

This is an example:
```
Signature keyId="SzFeHrH7uq13zen25GD+p+t7XZi8Xlc843SEIxdCR9E=",created=1609935692,algorithm="ecdsa-sha256",headers="(request-target) date digest x-request-id",signature="MEUCIBPKLQGjjKhpVzdQyU0tqkCXW2IlegWF3ZPEcqZL6AeWAiEA8eemIvapqTUHHmOPvG+lb53zuEpyXm8/j164Zi7R3mA="
```

Is it a bug? **Should created only be added if the `(created)` is set?**

I can provide a test case if you'd like and would be happy to make a PR. 

p.s. The changes to the constuctors have made Signature a little more confusing (especially since it is used for a Template and the Signature itself). I'd be willing to try to improve it (keeping compatibility) if you would be interested?

**Comment by rmannibucau** [comment]
Hi @mjeffrey ,

I hit the same bug. Issue is that https://github.com/tomitribe/http-signatures-java/blob/3a84217890d9c7d93d42585c4c9d86225d69f4ff/src/main/java/org/tomitribe/auth/signatures/Signer.java#L99 always creates created/expires values whereas it should only be valued lazily when evaluated in this loop https://github.com/tomitribe/http-signatures-java/blob/3a84217890d9c7d93d42585c4c9d86225d69f4ff/src/main/java/org/tomitribe/auth/signatures/Signatures.java#L64 (lazily).

Another bug related to banks I hit is that created/expires key name can be between parenthesis or not and this is not configurable.

If it helps, in my project code I moved from a signature with explicit values of all potential headers to a Function<String, String> to value the headers and timestamps which enables more easily to clean that up making computed fields lazy - indeed not using a lambda but class enables to bypass the Atomic stuff but it shares the idea:

```
    @Override
    public String compute(final String keyId, final PrivateKey key, final SigningAlgorithm algorithm,
                          final List<String> headers, final String method, final String pathAndQuery,
                          final Function<String, String> headerReader) {
        final var created = new AtomicLong();
        final var expires = new AtomicLong();
        final Function<String, String> findHeader = headerName -> {
            switch (headerName) {
                case "(request-target)":
                    return method.toLowerCase(ROOT) + ' ' + pathAndQuery;
                case "(created)":
                    if (created.get() == 0) {
                        created.set(Instant.now().getEpochSecond());
                    }
                    return Long.toString(created.get());
                case "(expires)":
                    if (created.get() == 0) {
                        created.set(Instant.now().getEpochSecond());
                    }
                    if (expires.get() == 0) {
                        expires.set(Instant.ofEpochSecond(created.get()).plusSeconds(60).getEpochSecond());
                    }
                    return Long.toString(expires.get());
                default:
                    return headerReader.apply(headerName);
            }
        };
        final String signingString = headers.stream()
                .map(header -> header + ": " + findHeader.apply(header))
                .collect(joining("\n"));
        final Signature signature;
        try {
            signature = Signature.getInstance(algorithm.getJvmAlgorithm());
            if (algorithm.getAlgorithmParameters() != null) {
                signature.setParameter(algorithm.getAlgorithmParameters());
            }
            signature.initSign(key);
            signature.update(signingString.getBytes(StandardCharsets.UTF_8));
            return "keyId=\"" + keyId + "\"," +
                    "algorithm=\"" + algorithm.getSigningAlgorithm() + "\"," +
                    "headers=\"" + String.join(" ", headers) + "\"," +
                    (created.get() > 0 ? "(created)=\"" + created + "\"," : "") +
                    (expires.get() > 0 ? "(expires)=\"" + expires + "\"," : "") +
                    "signature=\"" + Base64.getEncoder().encodeToString(signature.sign()) + "\"";
        } catch (final Exception e) {
            throw new IllegalArgumentException(e);
        }
    }
```

**Comment by mjeffrey** [comment]
Hi Romain

Thanks for the info. I guess the Banking issues you are referring to are the Berlin Group psd2 specifications. When those specs were made the signing spec did not include created or expired or hs2019 and so many Banks will not recognize them (and potentially fail). 

This puts a burden on the (clients) TPPs to adapt to each API (which is very much the case anyway since every api is different) .
so indeed, it is best to do what you do and explicitly create the final string yourself.

p.s. I'm on the Aspsp (bank) side so it is not so much of an issue for us, it was a unit test which fails in case of regression.

**Comment by dblevins** [comment]
Hit this myself.  I'll try to find time to ensure created isn't added unless needed.  Meanwhile, anyone encountering this can downgrade to v 1.4

**Comment by capslocky** [comment]
While this fix is yet to be released with version 1.8, we use the following dirty workaround.
```
PrivateKey key = PEM.readPrivateKey(new FileInputStream(privateKeyPath));
Signer signer = new Signer(key, signature);
Signature signed = signer.sign(method, uri, headers);
String authorization = signed.toString();

if(authorization.contains(",created=")){
	// this is a temp fix, needed only in the current actual lib version 1.7
	// it has been fixed in 1.8-SNAPSHOT but not published yet
	// https://github.com/tomitribe/http-signatures-java/commit/70189bb

	int index1 = authorization.indexOf(",created=");
	int index2 = authorization.indexOf(",", index1 + 1);

	authorization = authorization.substring(0, index1) + authorization.substring(index2);
}
```

**Comment by dblevins** [comment]
@capslocky Thanks for this ping.  Just released 1.8.  Let me know how it looks -- I don't recall if there was something unresolved that was holding up the release.

**Comment by dblevins** [comment]
@capslocky side note, have a look at https://github.com/tomitribe/churchkey for some significantly better key parsing support

**Comment by capslocky** [comment]
@dblevins wow, so fast! I have checked the newly released version 1.8, and it works just fine, without that dirty workaround.
Thank you!
